File: c/fiveproblems1.c
/* 
Write three functions that compute the sum of the numbers in a given list
using a for-loop, a while-loop, and recursion.
*/
#include <stdio.h>

int forloop(const int* buf, int len)
{
    int sum = 0;
    for (int i = 0; i < len; ++i)
        sum += buf[i];
    return sum;
}

int whileloop(const int* buf, int len)
{
    int sum = 0;
    while (len > 0)
        sum += buf[--len];
    return sum;
}

int recursion(const int* buf, int len)
{
    if (len == 0)
        return 0;
    --len;
    return buf[len] + recursion(buf, len);
}

int main(int argc, char* argv[])
{
    int buf[10] = {1,2,3,4,5,-4,-1,8,9,-17};
    printf("forloop(): %d\n", forloop(buf, 10));
    printf("whileloop(): %d\n", whileloop(buf, 10));
    printf("recursion(): %d\n", recursion(buf, 10));
    return 0;
}

File: rs/fiveproblems1.rs
// Write three functions that compute the sum of the numbers in a given list
// using a for-loop, a while-loop, and recursion.

fn main() {
    let l1: [i32; 10] = [0, 1, 1, 2, 3, 5, 8, 13, 21, 34];
    println!("forloop(): {}", forloop(&l1));
    println!("whileloop(): {}", whileloop(&l1));
    println!("recursion(): {}", recursion(&l1, 0));
    println!("sum(): {}", sum(&l1));
}

fn forloop(list: &[i32]) -> i32 {
    let mut sum = 0;
    for i in list.iter() {
        sum += i;
    }
    sum
}

fn whileloop(list: &[i32]) -> i32 {
    let mut sum = 0;
    let mut i = 0;
    while i < list.len() {
        sum += list[i];
        i += 1;
    }
    sum
}

fn recursion(list: &[i32], i: usize) -> i32 {
    if i >= list.len() {
        return 0;
    }
    return list[i] + recursion(list, i+1);
}

// Rust has nice collection methods
fn sum(list: &[i32]) -> i32 {
    list.iter().sum()
}
